### ISSUE TITLE
Pyright marks reportOptionalMemberAccess error for some instance attribute.

### ISSUE THREAD

**Issue by Rid1-fz-06** [microsoft/pyright#4365]
**Describe the bug**
Pyright marks ```reportOptionalMemberAccess``` error for some instance attribute even when all conditions are fine. Even the script runs without error and shows the correct result.

**To Reproduce**
1. Create a Node class with following instance variable
```python
value: Any
next_: Optional[Self]
prev: Optional[Self]
```
2. Add the following code segment to create ```Deque``` class
```python
class Deque:
    def __init__(self) -> None:
        self.head: Optional[Node] = None
        self.tail: Optional[Node] = None
        self.length = 0

    def append_right(self, value: Any) -> None:
        temp: Node = Node(value)
        if self.head is None:
            self.head = self.tail = temp
            self.length += 1
            return

        self.tail.next_ = temp
        temp.prev = self.tail
        self.length += 1
```
3. See the mark when ```self.tail.next_``` is accessed.

**Expected behavior**
It should not have caught and mark ```self.tail.next_``` as an error.

**Screenshots or Code**
Screenshot:
![image](https://user-images.githubusercontent.com/92540820/208845859-9131e709-f5a3-4294-b577-429643058698.png)

Code:
```python
class Node:
    def __init__(self, value: Any) -> None:
        self.value: Any = value
        self.prev: Optional[Any] = None
        self.next_: Optional[Any] = None


class Deque:
    def __init__(self) -> None:
        self.head: Optional[Node] = None
        self.tail: Optional[Node] = None
        self.length = 0

    def append_right(self, value: Any) -> None:
        temp: Node = Node(value)
        if self.head is None:
            self.head = self.tail = temp
            self.length += 1
            return

        self.tail.next_ = temp
        temp.prev = self.tail
        self.length += 1

```

**VS Code extension or command-line**
Pyright from within ```Pylance```
Pylance version: v2022.12.20
Pyright cli in ```neovim```
Pyright cli version: 1.1.284

**Comment by erictraut** [comment]
Pyright is correct in emitting an error here.

The type of `self.head` and `self.tail` are both `Any | None`.

In the `if` statement, you are explicitly checking to see if `self.head` is `None`. If this is not true`, execution proceeds after the `if` block, and the type of `self.head` is narrowed to `Any`. The type of `self.tail` is still `Any | None` because it has not been narrowed on this code path.

Your code happens to be structured such that `self.head` and `self.tail` are both `None` or both non-`None` values, but a static type checker isn't able to understand that assumption without some additional help.

My recommendation is that you add the statement `assert self.tail is not None` immediately before the statement `self.tail.next_ = temp`. This serves to document the assumption you're making in your code. It also validates this assumption at runtime and tells the type checker that it should assume `self.tail` can be narrowed on that code path.

For more information about type narrowing and type guards, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-narrowing).
